numba/_helperlib.c

```diff
@@ -1518,6 +1518,159 @@ numba_xxgemm(char kind, char *transa, char *transb,
     return 0;
 }
 
+/*
+ * LAPACK calling helpers.  The helpers can be called without the GIL held.
+ * The caller is responsible for checking arguments (especially dimensions).
+ */
+
+/* Fetch the address of the given CLAPACK function, as exposed by
+   scipy.linalg.cython_lapack. */
+static void *
+import_clapack_function(const char *function_name)
+{
+    PyObject *module, *capi, *cobj;
+    void *res = NULL;
+
+    module = PyImport_ImportModule("scipy.linalg.cython_lapack");
+    if (module == NULL)
+        return NULL;
+    capi = PyObject_GetAttrString(module, "__pyx_capi__");
+    Py_DECREF(module);
+    if (capi == NULL)
+        return NULL;
+    cobj = PyMapping_GetItemString(capi, function_name);
+    Py_DECREF(capi);
+    if (cobj == NULL)
+        return NULL;
+#if PY_MAJOR_VERSION >= 3 || (PY_MINOR_VERSION >= 7)
+    {
+        /* 2.7+ => Cython exports a PyCapsule */
+        const char *capsule_name = PyCapsule_GetName(cobj);
+        if (capsule_name != NULL) {
+            res = PyCapsule_GetPointer(cobj, capsule_name);
+        }
+        Py_DECREF(cobj);
+    }
+#else
+    {
+        /* 2.6 => Cython exports a legacy PyCObject */
+        res = PyCObject_AsVoidPtr(cobj);
+        Py_DECREF(cobj);
+    }
+#endif
+    return res;
+}
+
+/* Fast getters caching the value of a function's address after
+   the first call to import_clapack_function(). */
+
+#define EMIT_GET_CLAPACK_FUNC(name)                            \
+    static void *clapack_ ## name = NULL;                      \
+    static void *get_clapack_ ## name(void) {                  \
+        if (clapack_ ## name == NULL) {                        \
+            PyGILState_STATE st = PyGILState_Ensure();         \
+            clapack_ ## name = import_clapack_function(# name); \
+            PyGILState_Release(st);                            \
+        }                                                      \
+        return clapack_ ## name;                               \
+    }
+
+EMIT_GET_CLAPACK_FUNC(sgetrf)
+EMIT_GET_CLAPACK_FUNC(dgetrf)
+EMIT_GET_CLAPACK_FUNC(cgetrf)
+EMIT_GET_CLAPACK_FUNC(zgetrf)
+EMIT_GET_CLAPACK_FUNC(sgetri)
+EMIT_GET_CLAPACK_FUNC(dgetri)
+EMIT_GET_CLAPACK_FUNC(cgetri)
+EMIT_GET_CLAPACK_FUNC(zgetri)
+
+
+#undef EMIT_GET_CLAPACK_FUNC
+
+typedef void (*xxgetrf_t)(int *m, int *n, void *a, int *lda, int *ipiv, int *info);
+
+typedef void (*xxgetri_t)(int *n, void *a, int *lda, int *ipiv, void *work, int *lwork, int*info);
+
+/* Compute LU decomposition of a*/
+NUMBA_EXPORT_FUNC(int)
+numba_xxgetrf(char kind, Py_ssize_t m, Py_ssize_t n, void *a, Py_ssize_t lda, int *ipiv, int *info)  /* XXX should these be int * or Pyssize_t */
```

Comment:
I can't really comment on the correctness of these wrapper, perhaps @stuartarchibald wants to take a look?